source/fab/tasks/c.py

```diff
@@ -0,0 +1,535 @@
+            if node.kind == clang.cindex.CursorKind.FUNCTION_DECL:
+                if (node.is_definition()
+                        and node.linkage == clang.cindex.LinkageKind.EXTERNAL):
+                    # This should catch function definitions which are exposed
+                    # to the rest of the application
+                    current_def = CSymbolID(node.spelling, artifact.location)
+                    state.add_c_symbol(current_def)
+                    new_artifact.add_definition(node.spelling)
+                else:
+                    # Any other declarations should be coming in via headers,
+                    # we can use the injected pragmas to work out whether these
+                    # are coming from system headers or user headers
+                    if (self._check_for_include(node.location.line)
+                            == "usr_include"):
+                        usr_includes.append(node.spelling)
+
+            elif (node.kind == clang.cindex.CursorKind.CALL_EXPR):
+                # When encountering a function call we should be able to
+                # cross-reference it with a definition seen earlier; and
+                # if it came from a user supplied header then we will
+                # consider it a dependency within the project
+                if node.spelling in usr_includes and current_def is not None:
+                    # TODO: Assumption that the most recent exposed
+                    # definition encountered above is the one which
+                    # should lodge this dependency - is that true?
+                    state.add_c_dependency(current_def, node.spelling)
+                    new_artifact.add_dependency(node.spelling)
+
+        return [new_artifact]
+
+
+class _CTextReaderPragmas(TextReaderDecorator):
+    """
+    Reads a C source file but when encountering an #include
+    preprocessor directive injects a special Fab-specific
+    #pragma which can be picked up later by the Analyser
+    after the preprocessing
+    """
+    def __init__(self, source: TextReader):
+        super().__init__(source)
+        self._line_buffer = ''
+
+    _include_re: str = r'^\s*#include\s+(\S+)'
+    _include_pattern: Pattern = re.compile(_include_re)
+
+    def line_by_line(self) -> Iterator[str]:
+        for line in self._source.line_by_line():
+            include_match: Optional[Match] \
+                = self._include_pattern.match(line)
+            if include_match:
+                # For valid C the first character of the matched
+                # part of the group will indicate whether this is
+                # a system library include or a user include
+                include: str = include_match.group(1)
+                # TODO: Is this sufficient?  Or do the pragmas
+                #       need to include identifying info
+                #       e.g. the name of the original include?
+                if include.startswith('<'):
```

Comment:
We may need a less subtle solution. According to the standard the angle brackets are used for system headers and the quotes for non-system headers. However 3rd party headers can be included in such a way as to be accessed with angle brackets. More importantly 3rd party libraries like NetCDF will be accessed using quoted strings as they are not system headers. But they are library headers.

Reply:
I _think_ that's okay - in the same way as FCM if we have external libraries as part of the build then the configuration should contain that information (in an as yet unwritten detail) - so somewhere in the logic (the Analyser I suppose) will eventually be able to say _"this user header is on the list of external libraries, so I should treat things defined from it in a similar way to system headers"_ (i.e. by not trying to analyse dependencies in the usual way)?

Reply:
So long as it has been thought through I'm happy.